Only fire input value change events when the value changes

jquense opened the pull request:
More aggressive alternative to #5687, This PR tracks the value for all inputs and only fires a change when the value changed. This is similar in theory to how the onInput polyfill works __except__ it doesn't track the value in `activeElement` but on the input DOM node. It certainly starts to edge into "not to spec" territory, but onChange is already pretty unique in React so perhaps that's not important.

The extra state tracking isn't ideal, but it has benefits over the track-when-focused approach. For one testing is easier. detached inputs will still fire onChange events, which will break a lot less existing tests. I could also track the value on the node instance, instead of the DOM node, but I wasn't sure if how ya'll feel about mutating the component instance that way.

The downside to this approach is that manually triggered events won't fire if the value hasn't changed. As far as I can tell there isn't a way to distinguish between naturally triggered events and manually triggered ones. Not sure if it's worth it, since most are triggered in testing by the ReactTestUtils, it might make sense add a flag that way?

fixes #554, fixes #1471, fixes #2185 (that last one is probably caused by something else entirely but it was fixed in my testing)

cc: @jimfb @spicyj @syranide

jquense commented on src/renderers/dom/client/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -146,15 +177,13 @@ var isInputEventSupported = false;
-  // changes or when an input with a placeholder is focused.
   isInputEventSupported = isEventSupported('input') && (
-    !('documentMode' in document) || document.documentMode > 11
+    !('documentMode' in document) || document.documentMode > 9
```
reverts #4051 since this addresses that issue as well. Lets the polyfill be more focused since it has some downsides.

jquense commented on src/renderers/dom/client/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -45,6 +47,35 @@ var eventTypes = {
+  var value = getInputValue(target);
+  var lastValue = target[LAST_VALUE_KEY];
+
+  if (!target.hasOwnProperty(LAST_VALUE_KEY) || value !== lastValue) {
```
Not sure that tracking on the DOM node is the best option, perhaps the instance is better? I am not sure with object is more long-lived

jimfb replied:
By definition, they should both have the same lifespan.

jquense replied:
that was what I figured, my only thought was maybe nodes are destroyed/recreated for the same instance, but that didn't seem likely.

jimfb commented:
@spicyj I know you had some thoughts on this topic, just wanted to raise this PR for your attention/feedback.  It's nearly NYE, so maybe we won't get to this reviewed before the new year, but I do think we should fast track this PR to the extent possible :).

syranide commented on src/renderers/dom/client/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -45,6 +47,35 @@ var eventTypes = {
+  var lastValue = target[LAST_VALUE_KEY];
+
+  if (!target.hasOwnProperty(LAST_VALUE_KEY) || value !== lastValue) {
+    target[LAST_VALUE_KEY] = value;
```
We can't really do this, updating the value of inputs via script is allowed for uncontrolled inputs and these updates would not update `[LAST_VALUE_KEY]`. Also, as far as I can tell, this implementation does not update `[LAST_VALUE_KEY]` when the value an input is updated for controlled inputs either, but that is fixable (although slightly hacky perhaps).

jquense replied:
changing a value on on uncontrolled input should trigger an `onChange` which will trigger an update here (in text inputs anyway) since the code listens for both onChange and onInput. Missing an update should be fine since the next raised event will update the cached value 

That being said is that what we want in the first place? native `onInput` doesn't trigger when you set `node.value`; the IE8 pollyfill even wraps the `value` descriptor in order to detect and silence value changes triggered from javascript. If we need that consistenty (which makes sense to me) then wrapping the value descriptor in the same way would allow for that, but I'm not sure if that causes any significant perf regressions, getting and setting all value descriptors on listened to inputs.

jquense replied:
actually...it seems likes `node.value = 'foo'` doesn't trigger a change...which I could have sworn was not the case. Maybe there isn't any issue (from my perspective)

syranide replied:
> Missing an update should be fine since the next raised event will update the cached value

But it's not fine if you revert what the user typed and he types it again, it will then think the value hasn't changed, while uncommon that would be a severe bug.

I'm also skeptical of something like this if there are edge-cases where we are *still* triggering when the value hasn't actually changed (which can happen if the cache is wrong). If it's not 100% then the end benefit of this seems little more than a performance optimization, IMHO a good thing but perhaps not in relation to there being a fragile element to the implementation.

Also consider that we still don't know when auto-filling occurs in some browsers, this is a troublesome area.

jquense replied:
sorry slow this morning: yes failing test case

```js
it.only('should not fire change when setting the value programmatically', function() {
    var called = 0;

    function cb(e) {
      called += 1;
      expect(e.type).toBe('change');
    }

    var input = ReactTestUtils.renderIntoDocument(
      <input type="text" onChange={cb} defaultValue="foo"/>
    );

    ReactTestUtils.SimulateNative.change(input);
    input.value = 'bar'
    input.value = 'foo'
    ReactTestUtils.SimulateNative.change(input);

    expect(called).toBe(2);
  });
```

jquense commented:
I agree that _failing_ to report a change is not acceptable, I also think that over reporting in some cases is probably "ok" but most of the point of making this a sweeping change is to fix those sort of minor inconsistencies so not fixing them is sort of meh. 

That being said I think this approach is still salvageable.  I can't update the PR till later but I have some ideas that shouldn't be fragile...

jquense commented on src/renderers/dom/client/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -335,6 +378,43 @@ var ChangeEventPlugin = {
 
+  _isInputEventSupported: isInputEventSupported,
+
+  willDeleteListener(inst) {
```
I am not sure these hooks are appropriate for this sort of thing, but they seem like they are well suited for it, baring the below annoyance

jquense commented on src/renderers/dom/client/eventPlugins/ChangeEventPlugin.js:
```diff
@@ -335,6 +378,43 @@ var ChangeEventPlugin = {
+
+    // An uglier internal check to avoid a try/catch
+    // if the instance is unmounted the node will already be removed
+    if (inst._nativeNode !== null) {
```
When unmounting the node is already gone by the time this hook is called so the `getNodeFromInstance` call throws.

This seems like an unfortunate break in encapsulation, the alternative was to use a try/catch but I'm mostly conditioned at this point to be afraid of try/catch VM deopts.

Neo20067 replied:
$git--clone--repository--neo20067/https://github.com*/

jquense commented on src/renderers/dom/client/wrappers/__tests__/ReactDOMInput-test.js:
```diff
@@ -148,6 +151,7 @@ describe('ReactDOMInput', function() {
     var node = ReactDOM.render(stub, container);
 
     node.value = 'giraffe';
+    lastInputValue.set(node, undefined);
```
Leaky :/

jquense commented on src/renderers/dom/client/eventPlugins/__tests__/ChangeEventPlugin-test.js:
```diff
@@ -12,7 +12,18 @@
+
+var lastInputValue = ChangeEventPlugin.__lastInputValue;
+
+function setUntrackedValue(elem, value) {
```
Deduping the events and only firing when the value changes creates a new layer of event triggering concerns.

Generally in a test (or where ever) I think that the expected behavior of manually triggering an event is that that event is fired regardless of the underlying logic. Since we are creating a new boundry of when these events are "really" fired that diverges from the browser, we run into the issue of not being able to just let manually triggered events pass through. I think the least disruptive approach is to try and maintain that behavior. 

One solution would be flagging events in `simulateNativeEventOnNode` and then having the plugin check for the flag to determine if it should compare last and current values or just flush the event through. That feels kinda gross to me, but the current state of manually resetting the cache is also pretty damn ugly outside of the plugin and plugin tests.

jquense commented:
I'm not quite sure why some tests sporadically fail, this isn't happening "On My Machine"(tm). 

I would have thought setting `configurable` to true would stop that one. poking around some more

jimfb commented:
@spicyj @syranide Can someone with expertise in this area of the core take a look at this, as per https://github.com/facebook/react/pull/5746#issuecomment-167939675?  Thanks!

syranide commented:
https://github.com/facebook/react/pull/5746#discussion_r48592301 https://github.com/facebook/react/pull/5746#discussion_r48608800 are my thoughts on this.

I'd like to see this fixed, but unless you can 100% reliably track last value (which I don't think you can) then this seems dead in the water. We could kind of do it if we drop event bubbling (IMHO yes please) but we'd still have issues with external influences and AFAIK there is no event that reliably detects changes to the value across the board, even in modern browsers.

Please correct me if I'm wrong, but that's my understanding of the situation.

jquense commented:
@syranide is there any reason to think that spying on the value descriptor (along with the change/onInput events) will miss value updates since this is already the strategy we take for the polyfill? What strategy would you suggest to fix the noted issues?

I am happy to code whatever anyone wants to get these bugs fixed but it seems like the approaches I take are passed on, for the concern of edge cases. However the _current_ code is already full of these edge cases. At a certain point we are sticking with bugs for the fear of not covering every single edge case with a fix.  It seems the options are: 

- try and address each edge case directly
- solve the problem at a more holistic level

I've already tried the individual approaches to fixing each bug with an a fix unique to that bug, and those approaches were rejected for want of a more robust solution that dedupes the events.

I completely understand the concern that a change like this entails, and the can of worms it's potentially opening. I am happy to adjust the scope of the PR; If this seems like too much, then what about limiting it in scope as in: #5687 (needs to be updated to essentially this code but for ranges only) which at least solves one bad bug.

I am not attached to any of these solutions, I am attached to getting these bugs fixed, so if these approaches aren't helpful or acceptable, please recommend a different approach and I will be happy to do the leg work of getting it coded. Thanks again everyone

syranide commented:
> What strategy would you suggest to fix the noted issues?

@jquense Based on my (admittedly limited) understanding there is no 100% and reliable strategy for this problem, that's all I'm saying. AFAIK there are cases where we simply won't know the actual last value, that seems bad. I'm all for fixing this but *personally* if I have to choose between:

* Don't fix; inputs sometimes fire for noop.
* Fix; inputs will fire for noop in edge-cases, inputs will fail to fire for actual changes in edge-cases, the fix involves a lot of code.

That's my understanding of the situation right now and based on that "don't fix" seems like the best candidate. Again, please correct me if I'm wrong. But I'm happy for your contributions and ultimately it's up to the FB devs to decide, I'm not intimately familiar with the React event system.

jquense commented:
> Fix; inputs will fire for noop in edge-cases, inputs will fail to fire for actual changes in edge-cases, the fix involves a lot of code.

I can appreciate that, and I do want to make sure we are not missing any updates. However I want to be clear the above situation is _already_ the case for input ranges (#554), which is where the PR originally came from. Input events just do not fire on IE and versions of chrome (and i'm sure others) when the value changes in various situations.

This PR doesn't just try to address noop onInput calls (though that is a potentially nice side effect) it allows for listening for more events, in order to not miss changes.

Of course we could just not dedupe the extra events and live with that, I find that more acceptable then missing changes, which is to say I think we are on the same page in terms of trade offs. I am just saying we are already in the worst case option for some inputs.

syranide commented:
@jquense The other things you mention I have no issues with :+1:. I defer to @spicyj or someone "official", but personally I'm just against the fragile deduping logic.

> I am just saying we are already in the worst case option for some inputs.

True, but I feel like there is difference between such issues introduced by neglect and those introduced knowingly through added complexity.

jimfb commented:
As per conversation with Sebastian, going to tentatively accept.  We will sync with a few people internally to let them know that this change is coming down the pipeline, and push to www early in the week to maximize internal testing and make sure we're not breaking things.

@jquense can you fix the minor unit test failure?

@spicyj can you take a glance, and flag anything that you think would be a show stopper here?

Note to self: verify that we've cut v15 branch before merging.

spicyj commented:
It is hard for me to verify correctness here without a lot of thought, but I think this general strategy is okay. It would be better if we could refactor things to a point where ReactDOMInput could be responsible for tracking a lot of this data and logic instead of tacking on expandos in ChangeEventPlugin but that is probably significantly more difficult.

jquense commented:
@spicyj I don't think i'd be too tough to put the value tracking stuff in ReactDOMInput, it'd certainly remove some awkward tracking in the Event Plugin, I wasn't sure if it would be weird for the event plugin to reach out for that though. Perhaps ReactDOMInput could just expose utils for interacting with the tracking info.

Does that make more sense, should I also update the PR to handle that?

Also I'd like to subsume the old IE tracking logic so that tehre isn't too much confusing to follow duplication

jquense commented:
Ok @jimfb provided I didn't miss anything and the CI goes through we should be good.

### Updates
1. Moved value tracking logic to a seperate module and added a bunch more tests
2. Input tracking is stored on `inst._wrapperState` instead of the DOM node
3. Tracking setup and teardown is done in ReactDOMComponent for textareas and inputs, instead of the event plugin listener hooks.
4. fixed broken test.

Let me know if those changes make sense, if not I can revert and just fix the unit test.

jquense commented:
also confirmed that the old-ie polyfill still works on IE9

gaearon commented:
Putting a 15.x milestone here. We should review this ASAP after 15.0 is out so this doesn’t get stale.

jquense commented:
I think this might need to go on a major bump, as the change in behavior will break tests that manually trigger (native) change events on inputs, as is.  Also while folks shouldn't be depending on over reporting of change events (as in the radio button cases) this will break anyone who is. up to y'all tho. (the test breakage is avoidable but needs specific integration with .simulateNative)

jquense commented:
no chance of just getting this in 15.0.0? been a long time coming :P Id hate to watch it go stale again over the next bunch of months

spicyj commented:
We try not to pick stuff except for regressions after an RC, sorry.

jquense commented:
thanks for the heads up, I realize timing is the way it is, but this was accepted before the RC, wondering why it didn't get picked in this case? Sorry I had just been hopeful that after 10 (including the older PR) months this was gonna happen, now its another 4-6 months away.

gaearon commented:
Whatever the conclusion is I think we should find a way to ensure this does not get stale. For example we could have a `next` branch where this gets merged before master, that is synced with master every week.

jimfb commented:
Way back when this was accepted, I had a "Note to self: verify that we've cut v15 branch before merging.".  Sorry if that was a bit unclear, this was intended to land after v15.0.0.

gaearon commented:
>Note to self: verify that we've cut v15 branch before merging.

We still can’t merge in master, can we? Because presumably this would need to go into 16 while the master would be used for 15.x.

jimfb commented:
@gaearon No idea, I don't we've figured out how things are going to work under the new system.  My preference would be to merge into master and cherry-pick for the point releases, like we've always done historically.  Otherwise, where does new code (with potentially breaking changes) go?  We can't freeze all development until we think we're ready to do another major release.

gaearon commented:
>My preference would be to merge into master and cherry-pick for the point releases, like we've always done historically.

:+1: Thanks for clarifying! I haven’t been watching closely how this happens so I didn’t know that.

jquense commented:
ya'll can stick this in whatever release makes sense (despite my sadness :P) I would appreciate if we could merge promptly after ya'll have a safe `15.0.0` branch just b/c I'm not sure i'll have teh bandwidth to keep this updated into the immediate future.

thanks again everyone.

jimfb commented:
We've cut the branch, let's get this in.

jimfb commented:
Thanks @jquense!

jquense commented:
Woohoo! thanks y'all!

jimfb commented:
@jquense We're getting "descriptor.get.call is not a thing" errors on Safari that looks like it blames to this diff.  Can you take a look?

jquense commented:
as the recent recipient of a new mac book pro I can, for the firsts time say: "yes" to that :P I'll take a look over the weekend

jimfb commented:
Ping @jquense

jquense commented:
sorry new job, not a lot of free time, hope to try and debug today. we'll see!

jquense commented:
Quick test, seems to indicate that Safari returns `{ set: undefined, get: undefined, ...}` for the value descriptor of  the input...The issue is we need to wrap that to catch all value updates.

I am not very familiar with Safari, and can't find anything online about this, I don't know if its safe to assume that they are actually empty, or if Safari is hiding them. If they aren't there IU think we can jsut use our local copy of the value?

arendjr commented:
FYI: 15.0.x currently has a regression compared to 0.14.x with regards to auto-fill handling in IE11: #6614
This is a pretty big deal to us because our login form is currently pretty screwed up for people using auto-fill.

Now this PR should fix that, but I'm guessing you still merge this for 15.0.3... So does anyone know when 16.0.0 is scheduled?

jquense commented:
FYI folks, this could go into 15. and fix a bunch of regressions in IE related to me naively trying to fix cross browser event bugs, (and perhaps cause a bunch more, who can say? browsers are the worst)

The only "breaking" bit of this is the use native.simulate which could be made non-breaking fairly straight forwardly by marshalling fake events via some flag. It's ugly but I think its the most ergonomic choice _anyway_, those bits could always be removed in 16 if its too ugly. @jimfb @gaearon @zpao

spicyj commented:
@jquense Sorry if I'm missing something obvious, but can you elaborate on what the breaking bit is here?

jquense commented:
@spicyj `simulateNative` no longer works for `onChange` in the way it's generally used. 

Events only fire if the value in the event hasn't been seen before. so manually setting the value on an input and then calling `simulateNative` won't trigger a change. 

The simple enough fix is to flag events created by simulateNative, so that they pass through the event plugin;

see: https://github.com/facebook/react/pull/5746/files#diff-2bead3a73555ca476b3dedf1f31fbb93R108

ljharb commented:
Posted a comment [here](https://github.com/facebook/react/issues/1471#issuecomment-321705928) asking why this breaking change went into a minor version.

mathiasbynens commented:
In the Speedometer benchmark we used to trigger React’s `onChange` programmatically as follows:

```js
element.dispatchEvent(new Event('change', {
  bubbles: true,
  cancelable: true
}));
```

This patch [intentionally](https://github.com/facebook/react/commit/045f1a791c6e17253e9d927ffca70ae5d00b4fe5#diff-2bead3a73555ca476b3dedf1f31fbb93R68) broke such use cases. Is there a reasonable workaround?

sebmarkbage commented:
The core of that change is that the "internal" `value` must actually change for the event to trigger.

That's why these tests use internals to modify it: https://github.com/facebook/react/pull/5746/files#diff-2bead3a73555ca476b3dedf1f31fbb93

Before issuing the event.

You can also use a helper in ReactTestUtils to simulate an event happening at a higher level. `ReactTestUtils.Simulate.change(node)` but I'm not sure that works neither.

If you don't want any dependencies I think you can use the native setter to change the value before calling `dispatchEvent`.

```js
Object.getOwnPropertyDescriptor(HTMLInputElement.prototype, 'value').set.call(input, 'new value')
```

To ensure that the value has changed in the internals without us observing it.
